Util unit tests

erler opened the pull request:
Little warm-up PR for #150 

- removed deprecated test from storybook
- installed ts-jest
- added `util` tests for `blockToTime`, `formatBnBalance`, `getEncodedAddress` and `validation` (which are the ones I think it'd make sense to add tests)

Going forward it might make sense to add a `__mocks__` folder to the root directory for recurring mocks like `getNetwork`.

erler commented on front-end/src/util/__tests__/blockToTime.test.ts:
```diff
@@ -0,0 +1,15 @@
+import blockToTime from '../blockToTime';
+
+test('blocknumber to time format', () => {
+	const time1 = '1d 12h 40m ';
```
Note that this test would fail without a trailing space at the end. (Something to fix in the `blockToTime` script?)

Tbaut replied:
yup

erler commented on front-end/src/util/__tests__/formatBnBalance.test.ts:
```diff
@@ -0,0 +1,38 @@
+		jest.resetModules();
+	});
+
+	const formattedBalance1 = '1.00 ';
```
This would also fail without trailing space

Tbaut replied:
Same, this should be fixed.

Tbaut commented on front-end/src/util/__tests__/formatBnBalance.test.ts:
```diff
@@ -0,0 +1,38 @@
+	const formattedBalance3 = '53.0 KSM';
+	const formattedBalance4 = '53.0 DOT';
+
+	it('in Kusama', () => {
```
What does it do? We should have a better description like `it('verifies xyz in Kusama')`

erler replied:
I have another `describe` on line 10 (`Get Encoded Address`) -> so you get `Get Encoded Address in Kusama`

Tbaut commented on front-end/src/util/__tests__/getEncodedAddress.test.ts:
```diff
@@ -0,0 +1,28 @@
+
+	const address1 = 'FqfW96FuXRH7pRy9njeNSxu4CNoBM1hKnVAmNRrFrU2izwj';
+
+	it('for Kusama', () => {
```
Same here, when we have many tests, and some are failing, we need to understand right away what is failing.

erler commented:
Got rid off the trailing spaces in `blockToTime` and `formatBnBalance` and adjusted tests.

For `getEncodedAddress` and `formatBnBalance` I have a description in a `describe()` and then it takes the network from `it()` resulting in this message:
<img width="459" alt="Screenshot 2020-05-12 at 16 59 11" src="https://user-images.githubusercontent.com/7072141/81714545-77ca7f00-9477-11ea-931b-47d8fa4ce939.png">

Tbaut commented on front-end/src/util/__tests__/blockToTime.test.ts:
```diff
@@ -0,0 +1,15 @@
+// Copyright 2019-2020 @paritytech/polkassembly authors & contributors
+// This software may be modified and distributed under the terms
+// of the Apache-2.0 license. See the LICENSE file for details.
+import blockToTime from '../blockToTime';
```
Any reason not to use the same structure as for `formatBnBalance`?
These tests are quite hard to read IMO.
It'd be much easier to read something like:
it ('test blocktotime with and without blocktime set (default to polkadot)')
- test 1
- test 2 
^ both returning the same and `jest.mock for getNetwork` returns 'polkadot'

it('tests that an amount of block lower than 1min returns 0min')<-- actually we should probably return 1min
- test 3

it('tests a time to block equal to 1min returns 1min').

Also to ease readability I'd go with `const SEC = 1000` and then instead of `6000` have `6*SEC`. I wouldn't put the time2 and 3 as variables either. why not for time1 because it's reused but time2 and 3 aren't.

Tbaut commented on front-end/src/util/__tests__/formatBnBalance.test.ts:
```diff
@@ -0,0 +1,38 @@
+		getNetwork.mockImplementation(() => 'kusama');
+		expect(formatBnBalance('1000000000000', { numberAfterComma: 2 } )).toEqual(formattedBalance1);
+		expect(formatBnBalance('1000000000000000000', { numberAfterComma: 2 } )).not.toEqual(formattedBalance1);
+		expect(formatBnBalance('1000000000000000', { withThousandDelimitor: true } )).not.toEqual(formattedBalance2);
```
I don't understand this test in less than 2sec. `withThousanddelimitor` is true by default iirc. now we're testing that some value isn't equal to some other ?
If anything we should test `withThousandDelimitor: false`

Tbaut commented on front-end/src/util/__tests__/formatBnBalance.test.ts:
```diff
@@ -0,0 +1,38 @@
+	it('for Kusama', () => {
+		getNetwork.mockImplementation(() => 'kusama');
+		expect(formatBnBalance('1000000000000', { numberAfterComma: 2 } )).toEqual(formattedBalance1);
+		expect(formatBnBalance('1000000000000000000', { numberAfterComma: 2 } )).not.toEqual(formattedBalance1);
```
I don't think this shows anything tbh :)
```suggestion
```

Tbaut commented on front-end/src/util/__tests__/formatBnBalance.test.ts:
```diff
@@ -0,0 +1,38 @@
+		expect(formatBnBalance('1000000000000', { numberAfterComma: 2 } )).not.toEqual(formattedBalance1);
+		expect(formatBnBalance('1000000000000000000', { numberAfterComma: 2 } )).toEqual(formattedBalance1);
+		expect(formatBnBalance('1000000000000', { withThousandDelimitor: true } )).not.toEqual(formattedBalance2);
+		expect(formatBnBalance('1000000000000000000000', { numberAfterComma: 0, withThousandDelimitor: true } )).toEqual(formattedBalance2);
```
here again using variables really doesn't help readability. it doesn't matter if we repeat twice the same, at least it's easy to read.

Tbaut commented on front-end/src/util/__tests__/formatBnBalance.test.ts:
```diff
@@ -0,0 +1,38 @@
+		expect(formatBnBalance('1000000000000000000', { numberAfterComma: 2 } )).toEqual(formattedBalance1);
+		expect(formatBnBalance('1000000000000', { withThousandDelimitor: true } )).not.toEqual(formattedBalance2);
+		expect(formatBnBalance('1000000000000000000000', { numberAfterComma: 0, withThousandDelimitor: true } )).toEqual(formattedBalance2);
+		expect(formatBnBalance('53000000000000', { numberAfterComma: 1, withUnit: true } )).not.toEqual(formattedBalance3);
```
just like the others, this doesn't bring any added value.

Tbaut commented on front-end/src/util/__tests__/getEncodedAddress.test.ts:
```diff
@@ -0,0 +1,28 @@
+	it('for Kusama', () => {
+		getNetwork.mockImplementation(() => 'kusama');
+		expect(getEncodedAddress(address1)).toBe('FqfW96FuXRH7pRy9njeNSxu4CNoBM1hKnVAmNRrFrU2izwj');
+		expect(getEncodedAddress(address1)).not.toBe('A41F9YrXEYYxm2WFFrczYhUQiEW8APB8qw39gnYnzsYSZ9M');
```
test `FOO === BAR` and `FOO !== BAZ` is the same as `FOO === BAR` unless `FOO` is schizophrenic.
TL;DR, this second test doesn't bring any info.

edit: ah, now I understand why you have this second test. In this case, using variabled would make sense:
```js
const KUSAMA_ENCODED = 'Fq...'
const POLKADOT_ENCODET = 'A41..'

test for kusama:
-  KUSAMA_ENCODED => KUSAMA_ENCODED
-  POLKADOT_ENCODED => KUSAMA_ENCODED

test for polkadot 
POLKADOT_ENCODED => POLKADOT_ENCODED
KUSAMA_ENCODED => POLKADOT_ENCODED
```

Tbaut commented on front-end/src/util/__tests__/getEncodedAddress.test.ts:
```diff
@@ -0,0 +1,28 @@
+	it('for Polkadot', () => {
+		getNetwork.mockImplementation(() => 'polkadot');
+		expect(getEncodedAddress(address1)).not.toBe('FqfW96FuXRH7pRy9njeNSxu4CNoBM1hKnVAmNRrFrU2izwj');
+		expect(getEncodedAddress(address1)).toBe('A41F9YrXEYYxm2WFFrczYhUQiEW8APB8qw39gnYnzsYSZ9M');
```
dito

Tbaut commented on front-end/src/util/__tests__/validation.test.ts:
```diff
@@ -0,0 +1,43 @@
+	const fullname2 = 'Tim   Cook';
+	const fullname3 = 'X Æ A-12';
+
+	expect(fullname1.length).toBeGreaterThanOrEqual(fullname.minLength);
```
I don't understand this test, we're testing the name length? why?

Tbaut commented on front-end/src/util/__tests__/validation.test.ts:
```diff
@@ -0,0 +1,43 @@
+	const password1 = '123456';
+	const password2 = 'trailingSpace ';
+
+	expect(password1.length).toBeGreaterThanOrEqual(password.minLength);
```
Same here, why testing the length? We should test the validation, e.g what was done in test 1
it('should fail with a short password') -> test with "123"
it('should fail with a empty password') -> test with ""
it('should pass with a complex password') -> test with "^qr09@}= £$/"
...

erler replied:
Yeah, I get that.. But this is how `password` is defined in `validation.ts`
`export const password = { minLength: 6, required: true };`

There is no pattern property unlike with `email`

Tbaut replied:
ok then it doesn't make sense to test this. What needs to be tested (with enzyme) is the login component (which embeds this validation)

erler commented on front-end/src/util/blockToTime.ts:
```diff
@@ -9,11 +9,11 @@ function secondsToDhm(seconds: number) {
 	var d = Math.floor(seconds / (3600*24));
 	var h = Math.floor(seconds % (3600*24) / 3600);
-	var m = Math.floor(seconds % 3600 / 60);
+	var m = seconds < 60 ? Math.ceil(seconds % 3600 / 60) : Math.floor(seconds % 3600 / 60);
```
Added this so that for blocks equaling less than 1 min it'll get rounded up to 1min

Tbaut replied:
I guess the first expression will always = 1, so I suggest
```suggestion
	var m = seconds < 60 ? 1 : Math.floor(seconds % 3600 / 60);
```

erler commented:
- Removed most consts for better readability
- Added more descriptions
- Removed tests that are admittedly not telling much (e.g. for `password` and `fullname` in `validation`)
- adjusted the `blockToTime` script so that it'll round up blocks equaling to less than a minute to a full minute

More useful integration tests to be added with `Enzyme` or `react-testing-library`. This was just meant for `utils`.

erler commented on front-end/src/util/formatBnBalance.ts:
```diff
@@ -45,7 +45,7 @@ export  default function (value: BN | string, options: Options): string {
 		prefix = prefix.replace(/\B(?=(\d{3})+(?!\d))/g, ' ');
```
I'd be in favour of using a comma as a delimiter rather than a space. It's sort of the standard way of displaying balances.

Tbaut replied:
I'm strongly against that. It's standard in the US maybe, not in Europe.
Well Wikipedia seems to be ok with that: https://en.wikipedia.org/wiki/Decimal_separator#Digit_grouping

> Since 2003, the use of spaces as separators has been officially endorsed by SI/ISO 31-0 standard,[27] as well as by the International Bureau of Weights and Measures and the International Union of Pure and Applied Chemistry (IUPAC),[28][29] the American Medical Association's widely followed AMA Manual of Style, and the Metrication Board, among others.

This sums up my position, comma or points are misleading, and I still struggle with apps. I'd be even more in favor of a halfspace, or 0.3 rem, but this would require a little more work.
> Because of the confusion that could result in international documents, in recent years the use of spaces as separators has been advocated

erler replied:
Ok, this discussion will have to move to a general design consistency discussion. Just checked a range of crypto apps out there and pretty much everybody is using commas.
